utils/src/path_conversions.cpp

```diff
@@ -104,4 +105,142 @@ std::string path_to_uri(std::string_view path)
+    std::optional<std::string> fragment = std::nullopt;
+};
+
+void format_path_pre_processing(std::string& hostname, std::string& path)
+{
+    if (!hostname.empty() && !path.empty() && path[0] == '/')
+        path = path.substr(1, path.size() - 1);
+}
+
+void format_path_post_processing_win(std::string hostname, std::string& path)
+{
+    if (!path.empty() && hostname.empty())
+        path.erase(0, 1);
+}
+
+void format_path_post_processing_linux(std::string hostname, std::string& path)
+{
+    if (!hostname.empty())
+        path.insert(0, "/");
+}
+
+std::string format_path(std::string hostname, std::string path)
+{
+    format_path_pre_processing(hostname, path);
+
+    std::string s = utils::path::lexically_normal(utils::path::join(hostname, path)).string();
+
+    if (utils::platform::is_windows())
+        format_path_post_processing_win(hostname, s);
+    else
+        format_path_post_processing_linux(hostname, s);
+
+    return network::detail::decode(s);
+}
+
+std::string to_presentable_internal(const dissected_uri& dis_uri)
+{
+    std::string s;
+    if (dis_uri.scheme.has_value())
+        s.append(dis_uri.scheme.value()).append(":");
+    if (dis_uri.host.has_value())
+        s.append(dis_uri.host.value());
+    if (dis_uri.path.has_value())
+        s.append(dis_uri.path.value());
+
+    return s;
+}
+
+std::string to_presentable_internal_debug(const dissected_uri& dis_uri, std::string_view raw_uri)
+{
+    std::string s;
+    if (dis_uri.scheme.has_value())
+        s.append("Scheme: ").append(dis_uri.scheme.value()).append("\n");
+    if (dis_uri.auth.has_value())
+        s.append("Authority: ").append(dis_uri.auth.value()).append("\n");
+    if (dis_uri.path.has_value())
+        s.append("Path: ").append(dis_uri.path.value()).append("\n");
+    if (dis_uri.query.has_value())
+        s.append("Query: ").append(dis_uri.query.value()).append("\n");
+    if (dis_uri.fragment.has_value())
+        s.append("Fragment: ").append(dis_uri.fragment.value()).append("\n");
+
+    s.append("Raw URI: ").append(raw_uri);
+
+    return s;
+}
+
+dissected_uri dissect_uri(const std::string& uri, bool human_readable_only)
+{
+    dissected_uri ret;
+
+    try
+    {
+        network::uri u(uri);
+
+        if (u.has_scheme() && u.scheme().to_string() == "file")
+        {
+            std::string h;
+            std::string p;
+
+            if (u.has_host() && !u.host().empty())
+                h = u.host().to_string();
+            if (u.has_path())
+                p = u.path().to_string();
+            if (!h.empty() && u.has_authority())
+                h.insert(0, "//");
+
+            ret.path = format_path(h, p);
+
+            return ret;
+        }
+
+        if (u.has_scheme())
+            ret.scheme = u.scheme().to_string();
+        if (u.has_host())
+            ret.host = u.host().to_string();
+        if (ret.host.has_value() && u.has_authority())
+            ret.host->insert(0, "//");
```

Comment:
isn't this loosing e.g. port number? I think 3 cases need to be distinguished - no authority (untitled:filename), empty authority (file:///home) and the rest (file://server/filename, http://server:port/filename etc.)

Reply:
Query and fragment are also missing as you've mentioned below also . What is your opinion on printing them out for non_file URIs?